File: golang/internal/dedup/dedup.go
// Package dedup provides in-memory duplicate suppression, bounded by entry
// count and TTL (design §5 "Local state": bounded deduplication state). Not
// persistent — POC scope. Mirrors NotificationAgent.Core.Dedup.DeduplicationCache
// (C#) and notify_agent_core::dedup::DedupCache (Rust).
package dedup

import (
	"sync"
	"time"

	"github.com/YuShuanHsieh/os-notification/golang/internal/clock"
)

// entry pairs a key with the expiry time it was given at insertion. Kept in
// insertionOrder so that, combined with a fixed TTL and a monotonic clock,
// the queue stays expiry-ordered from front to back.
type entry struct {
	key    string
	expiry time.Time
}

// Cache is a bounded, TTL-based deduplication cache. It is safe for
// concurrent use by multiple goroutines.
type Cache struct {
	mu             sync.Mutex
	expiryByKey    map[string]time.Time
	insertionOrder []entry
	capacity       int
	ttl            time.Duration
	clk            clock.Clock
}

// NewCache returns a bounded, TTL-based dedup cache. capacity bounds the
// number of distinct keys retained (oldest/expired evicted first); ttl is how
// long a key continues to suppress duplicates after being added; clk is the
// time source.
func NewCache(capacity int, ttl time.Duration, clk clock.Clock) *Cache {
	if capacity < 1 {
		panic("dedup: capacity must be >= 1")
	}
	return &Cache{
		expiryByKey: make(map[string]time.Time, capacity),
		capacity:    capacity,
		ttl:         ttl,
		clk:         clk,
	}
}

// SeenOrAdd reports whether key was already present and not yet expired (a
// duplicate — the caller should drop the event without processing it
// further), or false if key is newly recorded (first-seen — caller proceeds
// normally). Safe for concurrent use from multiple goroutines with
// exactly-once "first seen" semantics per key even under a concurrent race.
func (c *Cache) SeenOrAdd(key string) bool {
	now := c.clk.Now()

	c.mu.Lock()
	defer c.mu.Unlock()

	// Purge expired entries from the front: fixed TTL + monotonic-ish clock
	// means the queue is expiry-ordered.
	for len(c.insertionOrder) > 0 && !c.insertionOrder[0].expiry.After(now) {
		c.dequeueOneLocked()
	}

	if existing, ok := c.expiryByKey[key]; ok && existing.After(now) {
		return true
	}

	for len(c.expiryByKey) >= c.capacity && len(c.insertionOrder) > 0 {
		c.dequeueOneLocked()
	}

	expires := now.Add(c.ttl)
	c.expiryByKey[key] = expires
	c.insertionOrder = append(c.insertionOrder, entry{key: key, expiry: expires})
	return false
}

// dequeueOneLocked removes the oldest entry from insertionOrder and, if it is
// still the current expiry for its key (a re-added key leaves a stale queue
// entry behind), removes it from expiryByKey too. Caller must hold c.mu.
func (c *Cache) dequeueOneLocked() {
	e := c.insertionOrder[0]
	c.insertionOrder = c.insertionOrder[1:]
	if current, ok := c.expiryByKey[e.key]; ok && current.Equal(e.expiry) {
		delete(c.expiryByKey, e.key)
	}
}

// size reports the current number of distinct keys held. Test-only helper.
func (c *Cache) size() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.expiryByKey)
}

File: rust/notify-agent-core/src/dedup.rs
use std::collections::{HashMap, VecDeque};
use std::sync::Mutex;
use tokio::time::{Duration, Instant};

/// In-memory duplicate suppression, bounded by entry count and TTL (design §5
/// "Local state"). Uses tokio::time::Instant so tests control the clock.
/// Not persistent — POC scope.
pub struct DedupCache {
    inner: Mutex<Inner>,
    capacity: usize,
    ttl: Duration,
}

struct Inner {
    expiry_by_key: HashMap<String, Instant>,
    insertion_order: VecDeque<(String, Instant)>,
}

impl DedupCache {
    pub fn new(capacity: usize, ttl: Duration) -> Self {
        assert!(capacity >= 1, "capacity must be >= 1");
        Self {
            inner: Mutex::new(Inner {
                expiry_by_key: HashMap::new(),
                insertion_order: VecDeque::new(),
            }),
            capacity,
            ttl,
        }
    }

    pub fn try_add(&self, key: &str) -> bool {
        let now = Instant::now();
        let mut g = self.inner.lock().unwrap();

        // Purge expired from the front: fixed TTL + monotonic clock means the
        // queue is expiry-ordered.
        while g.insertion_order.front().is_some_and(|(_, exp)| *exp <= now) {
            Self::dequeue_one(&mut g);
        }

        if g.expiry_by_key.get(key).is_some_and(|exp| *exp > now) {
            tracing::debug!(dedup_key = key, "dropping duplicate event (dedup key already seen within TTL)");
            return false;
        }

        while g.expiry_by_key.len() >= self.capacity && !g.insertion_order.is_empty() {
            Self::dequeue_one(&mut g);
        }

        let expires = now + self.ttl;
        g.expiry_by_key.insert(key.to_string(), expires);
        g.insertion_order.push_back((key.to_string(), expires));
        true
    }

    pub fn len(&self) -> usize {
        self.inner.lock().unwrap().expiry_by_key.len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    fn dequeue_one(g: &mut Inner) {
        if let Some((key, expires_at)) = g.insertion_order.pop_front() {
            // A re-added key leaves a stale queue entry behind; only remove on
            // exact expiry match.
            if g.expiry_by_key.get(&key) == Some(&expires_at) {
                g.expiry_by_key.remove(&key);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicU32, Ordering};
    use std::sync::Arc;
    use tokio::time::{advance, Duration};

    #[tokio::test(start_paused = true)]
    async fn first_add_true_second_false() {
        let cache = DedupCache::new(10, Duration::from_secs(600));
        assert!(cache.try_add("k1"));
        assert!(!cache.try_add("k1"));
        assert!(cache.try_add("k2"));
    }

    #[tokio::test(start_paused = true)]
    async fn key_expires_after_ttl() {
        let cache = DedupCache::new(10, Duration::from_secs(600));
        assert!(cache.try_add("k1"));
        advance(Duration::from_secs(540)).await; // 9 min: still within TTL
        assert!(!cache.try_add("k1"));
        advance(Duration::from_secs(120)).await; // 11 min since insert
        assert!(cache.try_add("k1"));
    }

    #[tokio::test(start_paused = true)]
    async fn evicts_oldest_when_over_capacity() {
        let cache = DedupCache::new(2, Duration::from_secs(3600));
        assert!(cache.try_add("a"));
        assert!(cache.try_add("b"));
        assert!(cache.try_add("c")); // evicts "a"
        assert!(cache.try_add("a")); // "a" was forgotten
        assert!(cache.len() <= 2);
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
    async fn concurrent_adds_have_exactly_one_winner_per_key() {
        let cache = Arc::new(DedupCache::new(10_000, Duration::from_secs(600)));
        let wins = Arc::new(AtomicU32::new(0));
        let mut handles = Vec::new();
        for i in 0..1000 {
            let cache = cache.clone();
            let wins = wins.clone();
            handles.push(tokio::spawn(async move {
                if cache.try_add("same-key") {
                    wins.fetch_add(1, Ordering::Relaxed);
                }
                cache.try_add(&format!("key-{i}"));
            }));
        }
        for h in handles {
            h.await.unwrap();
        }
        assert_eq!(wins.load(Ordering::Relaxed), 1);
    }
}
